Make "Proceed to Checkout" button non-sticky when zoom level is bigger than 100%

nielslange opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

When the screen width is smaller than 782px, the `Proceed to Checkout` button becomes sticky. When a user is zooming into the screen, the `Proceed to Checkout` button could overlap the product items. This PR prevents that the `Proceed to Checkout` button becomes sticky when the screen height is smaller than 500px, to ensure that the products items remain visible even with a high zoom factor.

Closes #43637.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add a product to the cart.
2. Zoom into the page to decrease the screen height.
3. Verify that the `Proceed to order` button is sticky when the screen height is bigger than 500px, and the screen width is smaller than 782px.
4. Verify that the `Proceed to order` button is no longer sticky when the screen height is 500px or less, and the screen width is smaller than 782px.

### Screenshots

<table>
<tr>
<td valign="top">
Zoom: 100% / Width: 600px / Height: 1000px
<br><br>
<img width="598" alt="Screenshot 2024-06-20 at 15 17 16" src="https://github.com/woocommerce/woocommerce/assets/3323310/48082102-663f-406d-9554-4ca2ac28f7dd">
</td>
<td valign="top">
Zoom: 200% / Width: 300px / Height: 500px
<br><br>
<img width="592" alt="Screenshot 2024-06-20 at 15 18 24" src="https://github.com/woocommerce/woocommerce/assets/3323310/fb6b99b8-67e3-4d4f-a86a-44289a023629">
</td>
</tr>
</table>

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

nielslange committed:
Make proceed to order button non sticky when zoom level is bigger than 100%

tarunvijwani commented on plugins/woocommerce-blocks/assets/js/blocks/cart/inner-blocks/proceed-to-checkout-block/block.tsx:
```diff
@@ -103,15 +103,18 @@ const Block = ( {
 			</div>
-			{ /* If the positionReferenceElement is below the viewport, display the sticky container. */ }
-			{ positionRelativeToViewport === 'below' && (
+			{ /* If the positionReferenceElement is below the viewport and if the the zoom level is <= 100%, display the sticky container. */ }
```
Since we're just adding a new class and style based on the conditions, I think we can simplify this code by using conditional classes with clsx.

Example: 

```js

	const zoomLevel = ( window.outerWidth / window.innerWidth ) * 100;
	const isStickyContainer =
		positionRelativeToViewport === 'below' && zoomLevel <= 100;

	const submitContainerClass = clsx( 'wc-block-cart__submit-container', {
		'wc-block-cart__submit-container--sticky': isStickyContainer,
	} );

	return (
		<div className={ clsx( 'wc-block-cart__submit', className ) }>
			{ positionReferenceElement }
			<div
				className={ submitContainerClass }
				style={ isStickyContainer ? { backgroundColor } : undefined }
			>
				{ submitContainerContents }
			</div>
		</div>
	);

```

nielslange replied:
Thanks for your suggestion, @tarunvijwani. As I've opted for a different approach, based on our discussion in p1718875553124459-slack-C8X6Q7XQU, I'd prefer to leave the logic as it is for now to prevent accidentally introducing a regression.

tarunvijwani reviewed (commented):
Thank you for working on this, Niels! The code looks good, but I'm unable to test the PR. The zoom level value remains at 100 all the time for me, even when I tried on a mobile device. Am I missing anything?

Testing site: https://prime-flute-raccoon.jurassic.ninja/


https://github.com/woocommerce/woocommerce/assets/11503784/dbd98d28-5115-4852-bc56-a3e22aceb88d

nielslange committed:
Define sticky container based on media query

nielslange committed:
Merge branch 'trunk' into update/43637-semi-sticky-proceed-to-checkout-button

nielslange committed:
Revert previous inline comment

nielslange committed:
Merge branch 'trunk' into update/43637-semi-sticky-proceed-to-checkout-button

nielslange committed:
Update display logic

nielslange commented:
Thanks for your review, @tarunvijwani. Based on our discussion in p1718875553124459-slack-C8X6Q7XQU, I've updated this PR. I would appreciate it if you could review the PR again.

tarunvijwani commented on plugins/woocommerce-blocks/assets/js/blocks/cart/inner-blocks/proceed-to-checkout-block/style.scss:
```diff
@@ -44,3 +38,15 @@
+	}
+}
+
+@media screen and (max-height: 500px) {
```
I think we should have min-height here. As it is now removing the sticky container for for mobile screens as well.

nielslange replied:
It's expected that the sticky button will disappear on mobile screens when the screen height is smaller than 500px, to ensure that the sticky button does not overlap the product line items. How would you use `min-height` in this case? This part is not clear to me.

tarunvijwani replied:
Please ignore my comment. I thought it was related to the screen orientation and maximum height, so I tested it on a mobile device as well.

tarunvijwani reviewed (changes requested):
Thank you for changing the approach. The changes are still not working for me. I also see a glitch, sticky container class gets added and removed at certain scroll height rapidly. 

https://github.com/woocommerce/woocommerce/assets/11503784/3f161150-152e-4bca-bcab-6eb64d8cf099

nielslange committed:
Adjust threshold to stabilise visibility toggling

nielslange commented:
> Thank you for changing the approach. The changes are still not working for me. I also see a glitch, sticky container class gets added and removed at certain scroll height rapidly.

Good catch, @tarunvijwani. Apparently, this issue was already existing before my PR, but went unnoticed until now. I just addressed it. Feel free to review the PR again.

tarunvijwani commented on plugins/woocommerce-blocks/assets/js/blocks/cart/inner-blocks/proceed-to-checkout-block/block.tsx:
```diff
@@ -106,18 +106,18 @@ const Block = ( {
 				>
 					{ submitContainerContents }
 				</div>
+			) : (
```
>Thanks for your suggestion, @tarunvijwani. As I've opted for a different approach, based on our discussion in p1718875553124459-slack-C8X6Q7XQU, I'd prefer to leave the logic as it is for now to prevent accidentally introducing a regression.

I still feel that using clsx we can make the code look simpler and easier to read. However, this is not a blocker, so I will approve the PR. 

```js
	// Determine if the container should be sticky based on the position relative to the viewport.
	const displayStickyContainer = positionRelativeToViewport === 'below';

	const submitContainerClass = clsx( 'wc-block-cart__submit-container', {
		'wc-block-cart__submit-container--sticky': displayStickyContainer,
	} );

	return (
		<div className={ clsx( 'wc-block-cart__submit', className ) }>
			{ positionReferenceElement }
			<div
				className={ submitContainerClass }
				style={ displayStickyContainer ? { backgroundColor }: {} }
			>
				{ submitContainerContents }
			</div>
		</div>
	);
	
```

nielslange replied:
I've applied your suggestion to increase the readability, @tarunvijwani. Thanks for suggesting this. 🙌

tarunvijwani reviewed (approved):
Thank you for the update @nielslange. It works great now.

nielslange committed:
Improve readability

nielslange committed:
Merge branch 'trunk' into update/43637-semi-sticky-proceed-to-checkout-button
